compiler/src/iree/compiler/Dialect/Stream/Transforms/UnifyEncodingForGlobals.cpp

```diff
@@ -0,0 +1,330 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Encoding/IR/EncodingDialect.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamTypes.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/Analysis/GlobalTable.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/DebugLog.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler::IREE::Stream {
+
+#define DEBUG_TYPE "iree-stream-unify-encoding-for-globals"
+
+#define GEN_PASS_DEF_UNIFYENCODINGFORGLOBALSPASS
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h.inc"
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Analysis.
+//===----------------------------------------------------------------------===//
+
+// Information about an encoded global and its relationship to its source.
+struct EncodedGlobalInfo {
+  IREE::Util::GlobalOpInterface encodedGlobal;
+  Attribute encodingAttr;
+  IREE::Stream::TensorEncodeOp encodeOp;
+  IREE::Stream::AffinityAttr affinityAttr;
+};
+
+// Information about a source global and all its encoded versions.
+struct SourceGlobalInfo {
+  IREE::Util::GlobalOpInterface sourceGlobal;
+  SmallVector<EncodedGlobalInfo> encodedVersions;
+};
+
+// Finds the global store op that uses the given value, traversing through
+// passthrough ops like clone.
+// TODO(#22485): Handle other passthrough ops (e.g., transfer).
+static IREE::Util::GlobalStoreOpInterface findStoreOp(Value value) {
+  for (auto user : value.getUsers()) {
```

Comment:
Do we expect the `tensor.encode` operation to have multiple users in the initializer?

If it has multiple users, how do we ensure that the first store is actually the store we're interested in?

Reply:
Similar question. Do we want to limit to the case where there is only one user?

Reply:
I may be over-designing the case when I evaluate control-flow support.. Let me strict it and we start with the IR chain we are seeing.